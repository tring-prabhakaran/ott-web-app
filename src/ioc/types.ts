export const SERVICES = {
  Account: Symbol.for('AccountService'),
  Api: Symbol.for('ApiService'),
  Checkout: Symbol.for('CheckoutService'),
  Cleeng: Symbol.for('CleengService'),
  Config: Symbol.for('ConfigService'),
  Entitlement: Symbol.for('EntitlementService'),
  EPG: Symbol.for('EPGService'),
  Favorites: Symbol.for('FavoritesService'),
  Subscription: Symbol.for('SubscriptionService'),
  WatchHistory: Symbol.for('WatchHistoryService'),
  Profile: Symbol.for('ProfileService'),
};

export const CONTROLLERS = {
  Account: Symbol.for('AccountController'),
  Api: Symbol.for('ApiController'),
  Checkout: Symbol.for('CheckoutController'),
  Config: Symbol.for('ConfigController'),
  Entitlement: Symbol.for('EntitlementController'),
  Favorites: Symbol.for('FavoritesController'),
  Settings: Symbol.for('SettingsController'),
  WatchHistory: Symbol.for('WatchHistoryController'),
  Epg: Symbol.for('EpgController'),
  Profile: Symbol.for('ProfileController'),
};

export const INTEGRATION_PROVIDER = {
  JW: 'JW',
  CLEENG: 'CLEENG',
} as const;

export const IS_INITIALIZED = 'IS_INITIALIZED';

export type ControllerType = keyof typeof CONTROLLERS;

export type IntegrationProvider = keyof typeof INTEGRATION_PROVIDER;
