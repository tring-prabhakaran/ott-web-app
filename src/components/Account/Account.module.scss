.textWithButtonContainer {
  display: flex;
  flex-direction: column;
  gap: 1rem;
}

.customFields {
  display: flex;
  flex-direction: column;
  gap: 0.5em;
}