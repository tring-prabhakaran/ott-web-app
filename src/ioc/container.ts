import { Container } from 'inversify';

const iocContainer = new Container({ defaultScope: 'Singleton' });

const getController = <T extends object>(type: symbol): T => {
  return iocContainer.get<T>(type);
};

export { iocContainer, getController };
