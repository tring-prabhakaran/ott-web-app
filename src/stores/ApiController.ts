import { inject, injectable } from 'inversify';

import type { GetPlaylistParams, Playlist, PlaylistItem } from '#types/playlist';
import { SERVICES } from '#src/ioc/types';
import type ApiService from '#src/services/api.service';
import type { AdSchedule } from '#types/ad-schedule';
import type { EpisodeInSeries, EpisodesWithPagination, GetSeriesParams, Series } from '#types/series';

@injectable()
export default class EpgController {
  private apiService: ApiService;

  constructor(@inject(SERVICES.Api) apiService: ApiService) {
    this.apiService = apiService;
  }

  async getPlaylistById(id?: string, params: GetPlaylistParams = {}): Promise<Playlist | undefined> {
    return this.apiService.getPlaylistById(id, params);
  }

  async getMediaByWatchlist(playlistId: string, mediaIds: string[], token?: string): Promise<PlaylistItem[] | undefined> {
    return this.apiService.getMediaByWatchlist(playlistId, mediaIds, token);
  }

  async getMediaById(id: string, token?: string, drmPolicyId?: string): Promise<PlaylistItem | undefined> {
    return this.apiService.getMediaById(id, token, drmPolicyId);
  }

  async getSeries(id: string, params: GetSeriesParams = {}): Promise<Series | undefined> {
    return this.apiService.getSeries(id, params);
  }

  async getSeriesByMediaIds(mediaIds: string[]): Promise<{ [mediaId: string]: EpisodeInSeries[] | undefined } | undefined> {
    return this.apiService.getSeriesByMediaIds(mediaIds);
  }

  async getEpisodes(args: { seriesId: string | undefined; pageOffset?: number; pageLimit?: number; afterId?: string }): Promise<EpisodesWithPagination> {
    return this.apiService.getEpisodes(args);
  }

  async getSeasonWithEpisodes(args: {
    seriesId: string | undefined;
    seasonNumber: number;
    pageOffset?: number;
    pageLimit?: number;
  }): Promise<EpisodesWithPagination> {
    return this.apiService.getSeasonWithEpisodes(args);
  }

  async getAdSchedule(id: string | undefined | null): Promise<AdSchedule | undefined> {
    return this.apiService.getAdSchedule(id);
  }
}
