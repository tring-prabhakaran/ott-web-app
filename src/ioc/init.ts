/* eslint-disable import/order */
// To organize imports in a better way

import { SERVICES, CONTROLLERS, IntegrationProvider, INTEGRATION_PROVIDER, IS_INITIALIZED } from './types';
import { iocContainer } from './container';

// Services
import ConfigService from '#src/services/config/config.service';
import EpgService from '#src/services/epg/epg.service';
import AccountCleengService from '#src/services/account/cleeng.account.service';
import AccountJWService from '#src/services/account/jw.account.service';
import CheckoutCleengService from '#src/services/checkout/cleeng.checkout.service';
import CheckoutJWService from '#src/services/checkout/jw.checkout.service';
import SubscriptionJWService from '#src/services/subscription/jw.subscription.service';
import SubscriptionCleengService from '#src/services/subscription/cleeng.subscription.service';
import CleengService from '#src/services/cleeng/cleeng.service';
import ApiService from '#src/services/api.service';
import EntitlementService from '#src/services/entitlement/entitlement.service';
import WatchHistoryService from '#src/services/watchHistory/watchhistory.service';
import FavoritesService from '#src/services/favorites/favorites.service';

// Controllers
import WatchHistoryController from '#src/stores/WatchHistoryController';
import AccountController from '#src/stores/AccountController';
import ConfigController from '#src/stores/ConfigController';
import FavoritesController from '#src/stores/FavoritesController';
import SettingsController from '#src/stores/SettingsController';
import EpgController from '#src/stores/EpgController';
import ApiController from '#src/stores/ApiController';
import EntitlementController from '#src/stores/EntitlementController';
import CheckoutController from '#src/stores/CheckoutController';

const CLEENG_SERVICES = {
  [SERVICES.Cleeng]: CleengService,
  [SERVICES.Account]: AccountCleengService,
  [SERVICES.Checkout]: CheckoutCleengService,
  [SERVICES.Subscription]: SubscriptionCleengService,
} as const;

const JW_SERVICES = {
  [SERVICES.Account]: AccountJWService,
  [SERVICES.Checkout]: CheckoutJWService,
  [SERVICES.Subscription]: SubscriptionJWService,
} as const;

const COMMON_CONTROLLERS = {
  [CONTROLLERS.WatchHistory]: WatchHistoryController,
  [CONTROLLERS.Favorites]: FavoritesController,
  [CONTROLLERS.Epg]: EpgController,
  [CONTROLLERS.Entitlement]: EntitlementController,
} as const;

const bindServices = (servicesMap: { [x: symbol]: new (...args: any[]) => object }) => {
  Object.getOwnPropertySymbols(servicesMap).forEach((key) => iocContainer.bind(key).to(servicesMap[key]));
};

export const initIOCData = (integration: IntegrationProvider | null) => {
  if (!iocContainer.isBound(IS_INITIALIZED)) {
    iocContainer.bind(IS_INITIALIZED).toConstantValue(true);
  }

  if (integration === INTEGRATION_PROVIDER.CLEENG) {
    bindServices(CLEENG_SERVICES);
  }

  if (integration === INTEGRATION_PROVIDER.JW) {
    bindServices(JW_SERVICES);
  }

  if (integration) {
    iocContainer.bind<AccountController>(CONTROLLERS.Account).to(AccountController);
    iocContainer.bind<CheckoutController>(CONTROLLERS.Checkout).to(CheckoutController);
  }

  bindServices(COMMON_CONTROLLERS);
};

// Config injectables initialized by default
// No need to reconfigure them on the config change
export const initDefaultServices = () => {
  if (iocContainer.isBound(IS_INITIALIZED)) {
    return;
  }

  iocContainer.bind<ApiService>(SERVICES.Api).to(ApiService);
  iocContainer.bind<EpgService>(SERVICES.EPG).to(EpgService);
  iocContainer.bind<ConfigService>(SERVICES.Config).to(ConfigService);
  iocContainer.bind<EntitlementService>(SERVICES.Entitlement).to(EntitlementService);
  iocContainer.bind<FavoritesService>(SERVICES.Favorites).to(FavoritesService);
  iocContainer.bind<WatchHistoryService>(SERVICES.WatchHistory).to(WatchHistoryService);

  iocContainer.bind<ConfigController>(CONTROLLERS.Config).to(ConfigController);
  iocContainer.bind<SettingsController>(CONTROLLERS.Settings).to(SettingsController);
  iocContainer.bind<ApiController>(CONTROLLERS.Api).to(ApiController);
};
