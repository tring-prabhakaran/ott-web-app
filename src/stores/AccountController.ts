import i18next from 'i18next';
import { inject, injectable, LazyServiceIdentifer } from 'inversify';

import type FavoritesController from './FavoritesController';
import type WatchHistoryController from './WatchHistoryController';

import { useConfigStore } from '#src/stores/ConfigStore';
import type {
  Capture,
  Customer,
  CustomerConsent,
  EmailConfirmPasswordInput,
  FirstLastNameInput,
  GetCaptureStatusResponse,
  GetCustomerConsentsResponse,
  GetPublisherConsentsResponse,
} from '#types/account';
import type { Offer } from '#types/checkout';
import { useAccountStore } from '#src/stores/AccountStore';
import { queryClient } from '#src/containers/QueryProvider/QueryProvider';
import { logDev } from '#src/utils/common';
import { CONTROLLERS, SERVICES } from '#src/ioc/types';
import type SubscriptionService from '#src/services/subscription.service';
import type AccountService from '#src/services/account.service';
import type CheckoutService from '#src/services/checkout.service';
import { useFavoritesStore } from '#src/stores/FavoritesStore';
import { useWatchHistoryStore } from '#src/stores/WatchHistoryStore';
import { addQueryParams } from '#src/utils/formatting';
import { simultaneousLoginWarningKey } from '#components/LoginForm/LoginForm';
import { ACCESS_MODEL } from '#src/config';

enum NotificationsTypes {
  ACCESS_GRANTED = 'access.granted',
  ACCESS_REVOKED = 'access.revoked',
  CARD_REQUIRES_ACTION = 'payment.card.requires.action',
  SUBSCRIBE_REQUIRES_ACTION = 'subscribe.requires.action',
  FAILED = '.failed',
  SUBSCRIBE_FAILED = 'subscribe.failed',
  ACCOUNT_LOGOUT = 'account.logout',
}

@injectable()
export default class AccountController {
  private checkoutService: CheckoutService;
  private accountService: AccountService;
  private subscriptionService: SubscriptionService;
  private favoritesController: FavoritesController;
  private watchHistoryController: WatchHistoryController;

  constructor(
    @inject(SERVICES.Checkout) checkoutService: CheckoutService,
    @inject(SERVICES.Account) accountService: AccountService,
    @inject(SERVICES.Subscription) subscriptionService: SubscriptionService,
    @inject(new LazyServiceIdentifer(() => CONTROLLERS.Favorites)) favoritesController: FavoritesController,
    @inject(new LazyServiceIdentifer(() => CONTROLLERS.WatchHistory)) watchHistoryController: WatchHistoryController,
  ) {
    this.checkoutService = checkoutService;
    this.accountService = accountService;
    this.subscriptionService = subscriptionService;
    this.favoritesController = favoritesController;
    this.watchHistoryController = watchHistoryController;
  }

  async initializeAccount() {
    const { config } = useConfigStore.getState();

    if (!this.accountService) {
      useAccountStore.setState({ loading: false });
      return;
    }

    const { canUpdateEmail, canRenewSubscription, canUpdatePaymentMethod, canChangePasswordWithOldPassword, canExportAccountData, canShowReceipts } =
      this.accountService;

    useAccountStore.setState({
      loading: true,
      canUpdateEmail,
      canRenewSubscription,
      canUpdatePaymentMethod,
      canChangePasswordWithOldPassword,
      canExportAccountData,
      canDeleteAccount: canExportAccountData,
      canShowReceipts,
    });

    await this.accountService.initialize(config, this.logout);

    try {
      const authData = await this.accountService.getAuthData();

      if (authData) {
        await this.getAccount();
        await this.watchHistoryController.restoreWatchHistory();
        await this.favoritesController.restoreFavorites();
      }
    } catch (error: unknown) {
      logDev('Failed to get user', error);

      // clear the session when the token was invalid
      // don't clear the session when the error is unknown (network hiccup or something similar)
      if (error instanceof Error && error.message.includes('Invalid JWT token')) {
        await this.logout();
      }
    }

    useAccountStore.setState({ loading: false });
  }

  async updatePersonalShelves() {
    const { watchHistory } = useWatchHistoryStore.getState();
    const { favorites } = useFavoritesStore.getState();
    const { getAccountInfo } = useAccountStore.getState();
    const { getSandbox } = useConfigStore.getState();

    const sandbox = getSandbox();
    const { customer } = getAccountInfo();

    if (!watchHistory && !favorites) return;

    const personalShelfData = {
      history: this.watchHistoryController.serializeWatchHistory(watchHistory),
      favorites: this.favoritesController.serializeFavorites(favorites),
    };

    return this.accountService?.updatePersonalShelves(
      {
        id: customer.id,
        externalData: personalShelfData,
      },
      sandbox,
    );
  }

  async updateUser(values: FirstLastNameInput | EmailConfirmPasswordInput): Promise<ServiceResponse<Customer>> {
    const { getSandbox } = useConfigStore.getState();
    const sandbox = getSandbox() || true;

    useAccountStore.setState({ loading: true });

    const { user, canUpdateEmail } = useAccountStore.getState();

    if (Object.prototype.hasOwnProperty.call(values, 'email') && !canUpdateEmail) {
      throw new Error('Email update not supported');
    }

    if (!user) {
      throw new Error('User not logged in');
    }

    const errors = this.validateInputLength(values as FirstLastNameInput);
    if (errors.length) {
      return {
        errors,
        responseData: {} as Customer,
      };
    }

    let payload = values;
    // this is needed as a fallback when the name is empty (cannot be empty on JWP integration)
    if (!this.accountService.canSupportEmptyFullName) {
      payload = { ...values, email: user.email };
    }

    const response = await this.accountService.updateCustomer({ ...payload, id: user.id.toString() }, sandbox);

    if (!response) {
      throw new Error('Unknown error');
    }

    if (response.errors?.length === 0) {
      useAccountStore.setState({ user: response.responseData });
    }

    return response;
  }

  async getAccount() {
    const { config, accessModel } = useConfigStore.getState();

    try {
      const response = await this.accountService.getUser({ config });
      if (response) {
        await this.afterLogin(response.user, response.customerConsents, accessModel);
      }

      useAccountStore.setState({ loading: false });
    } catch (error: unknown) {
      useAccountStore.setState({
        user: null,
        subscription: null,
        transactions: null,
        activePayment: null,
        customerConsents: null,
        publisherConsents: null,
        loading: false,
      });
    }
  }

  async login(email: string, password: string) {
    const { config, accessModel } = useConfigStore.getState();

    useAccountStore.setState({ loading: true });

    const response = await this.accountService.login({ config, email, password });
    if (response) {
      await this.afterLogin(response.user, response.customerConsents, accessModel);
      await this.favoritesController.restoreFavorites();
      await this.watchHistoryController.restoreWatchHistory();
    }

    useAccountStore.setState({ loading: false });
  }

  private async clearLoginState() {
    useAccountStore.setState({
      user: null,
      subscription: null,
      transactions: null,
      activePayment: null,
      customerConsents: null,
      publisherConsents: null,
      loading: false,
    });

    await this.favoritesController.restoreFavorites();
    await this.watchHistoryController.restoreWatchHistory();
  }

  async logout(logoutOptions: { includeNetworkRequest: boolean } = { includeNetworkRequest: true }) {
    // this invalidates all entitlements caches which makes the useEntitlement hook to verify the entitlements.
    await queryClient.invalidateQueries('entitlements');
    await this.clearLoginState();
    if (logoutOptions.includeNetworkRequest) {
      this.accountService?.logout();
    }
  }

  async register(email: string, password: string) {
    const { config, accessModel } = useConfigStore.getState();

    useAccountStore.setState({ loading: true });
    const response = await this.accountService.register({ config, email, password });
    if (response) {
      const { user, customerConsents } = response;
      await this.afterLogin(user, customerConsents, accessModel);
    }

    await this.updatePersonalShelves();
  }

  async updateConsents(customerConsents: CustomerConsent[]): Promise<ServiceResponse<CustomerConsent[]>> {
    const { getAccountInfo } = useAccountStore.getState();
    const { config } = useConfigStore.getState();
    const { customer } = getAccountInfo();

    useAccountStore.setState({ loading: true });

    try {
      const response = await this.accountService?.updateCustomerConsents({
        config,
        customer,
        consents: customerConsents,
      });

      if (response?.consents) {
        useAccountStore.setState({ customerConsents: response.consents });
        return {
          responseData: response.consents,
          errors: [],
        };
      }
      return {
        responseData: [],
        errors: [],
      };
    } finally {
      useAccountStore.setState({ loading: false });
    }
  }

  // TODO: Decide if it's worth keeping this or just leave combined with getUser
  // noinspection JSUnusedGlobalSymbols
  async getCustomerConsents(): Promise<GetCustomerConsentsResponse> {
    const { getAccountInfo } = useAccountStore.getState();
    const { config } = useConfigStore.getState();
    const { customer } = getAccountInfo();

    const response = await this.accountService.getCustomerConsents({ config, customer });

    if (response?.consents) {
      useAccountStore.setState({ customerConsents: response.consents });
    }

    return response;
  }

  async getPublisherConsents(): Promise<GetPublisherConsentsResponse> {
    const { config } = useConfigStore.getState();

    const response = await this.accountService.getPublisherConsents(config);

    useAccountStore.setState({ publisherConsents: response.consents });

    return response;
  }

  async getCaptureStatus(): Promise<GetCaptureStatusResponse> {
    const { getAccountInfo } = useAccountStore.getState();
    const { getSandbox } = useConfigStore.getState();
    const { customer } = getAccountInfo();
    const sandbox = getSandbox();

    const { responseData } = await this.accountService.getCaptureStatus({ customer }, sandbox);

    return responseData;
  }

  async updateCaptureAnswers(capture: Capture): Promise<Capture> {
    const { getAccountInfo } = useAccountStore.getState();
    const { getSandbox, accessModel } = useConfigStore.getState();
    const { customer, customerConsents } = getAccountInfo();
    const sandbox = getSandbox();

    const response = await this.accountService.updateCaptureAnswers({ customer, ...capture }, sandbox);

    if (response.errors.length > 0) throw new Error(response.errors[0]);

    await this.afterLogin(response.responseData as Customer, customerConsents, accessModel, false);

    return response.responseData;
  }

  async resetPassword(email: string, resetUrl: string) {
    const { getSandbox, getAuthProvider } = useConfigStore.getState();
    const sandbox = getSandbox();
    const authProviderId = getAuthProvider();

    const response = await this.accountService.resetPassword(
      {
        customerEmail: email,
        publisherId: authProviderId,
        resetUrl,
      },
      sandbox,
    );

    if (response.errors.length > 0) throw new Error(response.errors[0]);

    return response.responseData;
  }

  async changePasswordWithOldPassword(oldPassword: string, newPassword: string, newPasswordConfirmation: string) {
    const { getSandbox } = useConfigStore.getState();
    const sandbox = getSandbox();

    const response = await this.accountService.changePasswordWithOldPassword(
      {
        oldPassword,
        newPassword,
        newPasswordConfirmation,
      },
      sandbox,
    );
    if (response?.errors?.length > 0) throw new Error(response.errors[0]);

    return response?.responseData;
  }

  async changePasswordWithToken(customerEmail: string, newPassword: string, resetPasswordToken: string, newPasswordConfirmation: string) {
    const { getSandbox, getAuthProvider } = useConfigStore.getState();
    const sandbox = getSandbox();
    const authProviderId = getAuthProvider();

    const response = await this.accountService.changePasswordWithResetToken(
      { publisherId: authProviderId, customerEmail, newPassword, resetPasswordToken, newPasswordConfirmation },
      sandbox,
    );
    if (response?.errors?.length > 0) throw new Error(response.errors[0]);

    return response?.responseData;
  }

  async updateSubscription(status: 'active' | 'cancelled'): Promise<unknown> {
    const { getSandbox } = useConfigStore.getState();
    const { getAccountInfo } = useAccountStore.getState();

    const { customerId } = getAccountInfo();
    const sandbox = getSandbox();

    if (!this.subscriptionService) throw new Error('subscription service is not configured');
    const { subscription } = useAccountStore.getState();
    if (!subscription) throw new Error('user has no active subscription');

    const response = await this.subscriptionService?.updateSubscription(
      {
        customerId,
        offerId: subscription.offerId,
        status,
        unsubscribeUrl: subscription.unsubscribeUrl,
      },
      sandbox,
    );

    if (response.errors.length > 0) throw new Error(response.errors[0]);

    await this.reloadActiveSubscription({ delay: 2000 });

    return response?.responseData;
  }

  async updateCardDetails({
    cardName,
    cardNumber,
    cvc,
    expMonth,
    expYear,
    currency,
  }: {
    cardName: string;
    cardNumber: string;
    cvc: number;
    expMonth: number;
    expYear: number;
    currency: string;
  }) {
    const { getSandbox } = useConfigStore.getState();
    const { getAccountInfo } = useAccountStore.getState();

    const { customerId } = getAccountInfo();
    const sandbox = getSandbox();

    const response = await this.subscriptionService?.updateCardDetails({ cardName, cardNumber, cvc, expMonth, expYear, currency }, sandbox);
    const activePayment = (await this.subscriptionService?.getActivePayment({ sandbox, customerId })) || null;
    useAccountStore.setState({
      loading: false,
      activePayment,
    });
    return response;
  }

  async checkEntitlements(offerId?: string): Promise<unknown> {
    const { getSandbox } = useConfigStore.getState();

    const sandbox = getSandbox();

    if (!this.checkoutService) throw new Error('checkout service is not configured');
    if (!offerId) {
      return false;
    }
    const { responseData } = await this.checkoutService.getEntitlements({ offerId }, sandbox);
    return !!responseData?.accessGranted;
  }

  async reloadActiveSubscription({ delay }: { delay: number } = { delay: 0 }): Promise<unknown> {
    useAccountStore.setState({ loading: true });
    const { getSandbox, config } = useConfigStore.getState();
    const { getAccountInfo } = useAccountStore.getState();

    const { customerId } = getAccountInfo();
    const sandbox = getSandbox();

    if (!this.subscriptionService) throw new Error('subscription service is not configured');
    // The subscription data takes a few seconds to load after it's purchased,
    // so here's a delay mechanism to give it time to process
    if (delay > 0) {
      return new Promise((resolve: (value?: unknown) => void) => {
        window.setTimeout(() => {
          this.reloadActiveSubscription().finally(resolve);
        }, delay);
      });
    }

    const [activeSubscription, transactions, activePayment] = await Promise.all([
      this.subscriptionService.getActiveSubscription({ sandbox, customerId, config }),
      this.subscriptionService.getAllTransactions({ sandbox, customerId }),
      this.subscriptionService.getActivePayment({ sandbox, customerId }),
    ]);

    let pendingOffer: Offer | null = null;

    // resolve and fetch the pending offer after upgrade/downgrade
    try {
      if (activeSubscription?.pendingSwitchId && this.checkoutService && 'getSubscriptionSwitch' in this.checkoutService) {
        const switchOffer = await this.checkoutService.getSubscriptionSwitch({ switchId: activeSubscription.pendingSwitchId }, sandbox);
        const offerResponse = await this.checkoutService.getOffer({ offerId: switchOffer.responseData.toOfferId }, sandbox);

        pendingOffer = offerResponse.responseData;
      }
    } catch (error: unknown) {
      logDev('Failed to fetch the pending offer', error);
    }

    // this invalidates all entitlements caches which makes the useEntitlement hook to verify the entitlements.
    await queryClient.invalidateQueries('entitlements');

    useAccountStore.setState({
      subscription: activeSubscription,
      pendingOffer,
      loading: false,
      transactions,
      activePayment,
    });
  }

  async exportAccountData() {
    return this.accountService.exportAccountData(undefined, true);
  }

  async getSocialLoginUrls() {
    const { config } = useConfigStore.getState();

    return this.accountService.getSocialUrls(config);
  }

  async deleteAccountData(password: string) {
    return this.accountService.deleteAccount({ password }, true);
  }

  async getReceipt(transactionId: string) {
    const { getSandbox } = useConfigStore.getState();
    const sandbox = getSandbox();

    if (!this.subscriptionService || !('fetchReceipt' in this.subscriptionService)) return null;

    const { responseData } = await this.subscriptionService.fetchReceipt({ transactionId }, sandbox);

    return responseData;
  }

  private async afterLogin(user: Customer, customerConsents: CustomerConsent[] | null, accessModel: string, shouldSubscriptionReload: boolean = true) {
    useAccountStore.setState({
      user,
      customerConsents,
    });

    this.subscribeToNotifications(user.uuid);

    return await Promise.allSettled([
      accessModel === ACCESS_MODEL.SVOD && shouldSubscriptionReload ? this.reloadActiveSubscription() : Promise.resolve(),
      this.getPublisherConsents(),
    ]);
  }

  private validateInputLength = (values: { firstName: string; lastName: string }) => {
    const errors: string[] = [];
    if (Number(values?.firstName?.length) > 50) {
      errors.push(i18next.t('account:validation.first_name'));
    }
    if (Number(values?.lastName?.length) > 50) {
      errors.push(i18next.t('account:validation.last_name'));
    }

    return errors;
  };

  async subscribeToNotifications(uuid: string = '') {
    return await this.accountService.subscribeToNotifications({
      uuid,
      onMessage: async (message) => {
        if (message) {
          const notification = JSON.parse(message);
          switch (notification.type) {
            case NotificationsTypes.FAILED:
            case NotificationsTypes.SUBSCRIBE_FAILED:
              window.location.href = addQueryParams(window.location.href, { u: 'payment-error', message: notification.resource?.message });
              break;
            case NotificationsTypes.ACCESS_GRANTED:
              await this.reloadActiveSubscription();
              break;
            case NotificationsTypes.ACCESS_REVOKED:
              await this.reloadActiveSubscription();
              break;
            case NotificationsTypes.CARD_REQUIRES_ACTION:
            case NotificationsTypes.SUBSCRIBE_REQUIRES_ACTION:
              window.location.href = notification.resource?.redirect_to_url;
              break;
            case NotificationsTypes.ACCOUNT_LOGOUT:
              if (notification.resource?.reason === 'sessions_limit') {
                window.location.href = addQueryParams(window.location.href, { u: 'login', message: simultaneousLoginWarningKey });
              } else {
                await this.logout();
              }
              break;
            default:
              break;
          }
        }
      },
    });
  }

  getAuthData() {
    return this.accountService.getAuthData();
  }
}
