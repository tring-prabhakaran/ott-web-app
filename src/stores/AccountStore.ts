import { createStore } from '#src/stores/utils';
import type { Consent, Customer, CustomerConsent } from '#types/account';
import type { Offer } from '#types/checkout';
import type { PaymentDetail, Subscription, Transaction } from '#types/subscription';

type AccountStore = {
  loading: boolean;
  user: Customer | null;
  subscription: Subscription | null;
  transactions: Transaction[] | null;
  activePayment: PaymentDetail | null;
  customerConsents: CustomerConsent[] | null;
  publisherConsents: Consent[] | null;
  pendingOffer: Offer | null;
  canUpdateEmail: boolean;
  canRenewSubscription: boolean;
  canUpdatePaymentMethod: boolean;
  canChangePasswordWithOldPassword: boolean;
  canExportAccountData: boolean;
  canDeleteAccount: boolean;
  canShowReceipts: boolean;
  canManageProfiles: boolean;
  setLoading: (loading: boolean) => void;
  getAccountInfo: () => { customerId: string; customer: Customer; customerConsents: CustomerConsent[] | null };
};

export const useAccountStore = createStore<AccountStore>('AccountStore', (set, get) => ({
  loading: true,
  user: null,
  subscription: null,
  transactions: null,
  activePayment: null,
  customerConsents: null,
  publisherConsents: null,
  pendingOffer: null,
  canUpdateEmail: false,
  canRenewSubscription: false,
  canChangePasswordWithOldPassword: false,
  canExportAccountData: false,
  canDeleteAccount: false,
  canUpdatePaymentMethod: false,
  canShowReceipts: false,
  canManageProfiles: false,
  setLoading: (loading: boolean) => set({ loading }),
  getAccountInfo: () => {
    const user = get().user;
    const customerConsents = get().customerConsents;

    if (!user?.id) throw new Error('user not logged in');

    return { customerId: user?.id, customer: user, customerConsents };
  },
}));
