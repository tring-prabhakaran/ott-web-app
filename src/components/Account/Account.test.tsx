import React from 'react';

import Account from './Account';

import customer from '#test/fixtures/customer.json';
import { useAccountStore } from '#src/stores/AccountStore';
import { renderWithRouter } from '#test/testUtils';
import type { Consent } from '#types/account';
import { CONTROLLERS } from '#src/ioc/types';

vi.mock('#src/ioc/container', () => ({
  getController: (type: symbol) => {
    switch (type) {
      case CONTROLLERS.Account:
        return { exportAccountData: vi.fn() };
    }
  },
}));

describe('<Account>', () => {
  test('renders and matches snapshot', () => {
    useAccountStore.setState({
      user: customer,
      publisherConsents: Array.of({ name: 'marketing', label: 'Receive Marketing Emails' } as Consent),
    });

    const { container } = renderWithRouter(<Account panelClassName={'panel-class'} panelHeaderClassName={'header-class'} canUpdateEmail={true} />);

    // todo
    expect(container).toMatchSnapshot();
  });
});
