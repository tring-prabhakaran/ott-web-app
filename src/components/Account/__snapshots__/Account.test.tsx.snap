// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Account> > renders and matches snapshot 1`] = `
<div>
  <div
    class="panel-class"
  >
    <div
      class="header-class"
    >
      <h3>
        account.about_you
      </h3>
    </div>
    <form
      class="_flexBox_1c1c63"
      novalidate=""
    >
      <div
        class="_textField_e16c1b"
      >
        <label
          class="_label_e16c1b"
          for="text-field_1235_firstName"
        >
          account.firstname
        </label>
        <p>
          John
        </p>
      </div>
      <div
        class="_textField_e16c1b"
      >
        <label
          class="_label_e16c1b"
          for="text-field_1235_lastName"
        >
          account.lastname
        </label>
        <p>
          Doe
        </p>
      </div>
    </form>
    <div
      class="_controls_1c1c63"
    >
      <button
        class="_button_f8f296 _default_f8f296 _outlined_f8f296"
        type="button"
      >
        <span>
          account.edit_information
        </span>
      </button>
    </div>
  </div>
  <div
    class="panel-class"
  >
    <div
      class="header-class"
    >
      <h3>
        account.email
      </h3>
    </div>
    <form
      class="_flexBox_1c1c63"
      novalidate=""
    >
      <div
        class="_textField_e16c1b"
      >
        <label
          class="_label_e16c1b"
          for="text-field_1235_email"
        >
          account.email
        </label>
        <p>
          email@domain.com
        </p>
      </div>
    </form>
    <div
      class="_controls_1c1c63"
    >
      <button
        class="_button_f8f296 _default_f8f296 _outlined_f8f296"
        type="button"
      >
        <span>
          account.edit_account
        </span>
      </button>
    </div>
  </div>
  <div
    class="panel-class"
  >
    <div
      class="header-class"
    >
      <h3>
        account.security
      </h3>
    </div>
    <form
      class="_flexBox_1c1c63"
      novalidate=""
    />
    <div
      class="_controls_1c1c63"
    >
      <button
        class="_button_f8f296 _default_f8f296 _outlined_f8f296"
        type="button"
      >
        <span>
          account.edit_password
        </span>
      </button>
    </div>
  </div>
  <div
    class="panel-class"
  >
    <div
      class="header-class"
    >
      <h3>
        account.terms_and_tracking
      </h3>
    </div>
    <form
      class="_flexBox_1c1c63"
      novalidate=""
    />
    <div
      class="_controls_1c1c63"
    />
  </div>
  <div
    class="panel-class"
  >
    <div
      class="header-class"
    >
      <h3>
        account.other_registration_details
      </h3>
    </div>
    <form
      class="_flexBox_1c1c63"
      novalidate=""
    >
      <div
        class="_customFields_c968f7"
      >
        <div
          class="_checkbox_531f07"
        >
          <div
            class="_row_531f07"
          >
            <input
              id="check-box_1235_consentsValues.marketing"
              name="consentsValues.marketing"
              type="checkbox"
              value=""
            />
            <label
              for="check-box_1235_consentsValues.marketing"
            >
              Receive Marketing Emails
            </label>
          </div>
        </div>
      </div>
    </form>
    <div
      class="_controls_1c1c63"
    />
  </div>
</div>
`;
